src/provider/provider_tracking.c

```diff
@@ -124,24 +280,43 @@ umf_result_t umfMemoryTrackerGetAllocInfo(const void *ptr,
         return UMF_RESULT_ERROR_NOT_SUPPORTED;
     }
 
-    if (TRACKER->alloc_segments_map == NULL) {
-        LOG_ERR("tracker's alloc_segments_map does not exist");
-        return UMF_RESULT_ERROR_NOT_SUPPORTED;
+    for (int i = 0; i < MAX_LEVELS_OF_MAPS; i++) {
+        if (TRACKER->alloc_segments_map[i] == NULL) {
+            LOG_ERR("tracker's alloc_segments_map does not exist");
+            return UMF_RESULT_ERROR_NOT_SUPPORTED;
+        }
     }
 
-    uintptr_t rkey;
-    tracker_alloc_info_t *rvalue;
-    int found = critnib_find(TRACKER->alloc_segments_map, (uintptr_t)ptr,
+    tracker_alloc_info_t *last_value = NULL;
+    tracker_alloc_info_t *rvalue = NULL;
+    uintptr_t last_key = 0;
+    uintptr_t rkey = 0;
+    int level = 0;
+    int found = 0;
+
+    do {
```

Comment:
I see this loop a few times in the code. Could you create a function for it?

Reply:
Done. Only two different ones left.